### ISSUE TITLE
kivy-i18n integration with core / Pango text provider

### ISSUE THREAD

**Issue by ghost** [RevengeComing/kivy-i18n#1]
Hello! I've been working on a Pango text provider for Kivy, solving bidirectional text and font fallback are the main priorities, [see pull request 5320](https://github.com/kivy/kivy/pull/5320) for details. One big problem for me is that I don't know any RTL language, it's hard to figure out what the desired behavior should is in the first place, let alone know if something working or broken...

I am opening this issue mostly for discovering/discussing the overlap between our work. Since Kivy does its own text layout, we can't rely on Pango to handle bidirectional alignment. Currently I believe changes are needed to text layout to deal with this correctly, but I'm not sure how it's supposed to behave. This may overlap with the missing alignment in kivy-i18n, maybe it makes sense to pool resources here?

I could very much use input on the Pango PR in general. Currently I have no idea if it works correctly rendering RTL. If you have some free time, could you please test it and leave some feedback? Thanks in advance :)

**Comment by RevengeComing** [comment]
Hi @bionoid,
First of all I'm really happy to see you did this job with pango. after weeks working on this repository i found out pango was better choice since it is higher level than HarfBuzz+Fribidi becuase you need to create a layout ( or use kivy's layout and use HB and Fribidi's functions) but I had no time to test it.

> maybe it makes sense to pool resources here?

Sure, I will take a look at your PR and I will tell you what can be used from my works.

> If you have some free time, could you please test it and leave some feedback?

Sure , i will be glad if i can help on Kivy's bidirectional text rendering.

I will leave a comment here when i did it ( it might take some time(2-4weeks) since i'm busy right now ).

**Comment by ghost** [comment]
> after weeks working on this repository i found out pango was better choice

I am not sure that is really true :) hb+fribidi is likely simpler to build and distribute cross-platform, I think both approaches have merit. 

> becuase you need to create a layout ( or use kivy's layout and use HB and Fribidi's functions) 

Unfortunately this situation is the in Pango PR - yes, it uses Pango layouts, but not for actual text layout (still handled by Kivy). A new text layout provider abstraction could be added, so that Pango could be plugged into do the full layout. Or the layout mechanism could be completely sidestepped by providing `PangoLabel`, `PangoTextInput` etc, but that doesn't seem like a good solution.

> Sure, I will take a look at your PR and I will tell you what can be used from my works.

I've looked around in your repository, I don't think there is much direct overlap at the moment. I was thinking mostly that the unsolved layout problems will overlap, and your README suggested that you are planning to work on it :) Is there any reason you did not write this as a `text_pyharfbuzz` provider? I think that would be ideal, we could adapt the core text API so both Pango and HarfBuzz have what they need to provide RTL/bidi. That should be a solid foundation for the future I think!

In any case, I did some more work since opening this issue, now including alignment for Label. I haven't looked at TextInput yet, that's a big unknown at the moment. It probably needs to itemize the text, maybe use two cursors, etc. I will try to mock something up in the next week(s).

Thanks for your interest & looking forward to hear from you :)

**Comment by RevengeComing** [comment]
> Is there any reason you did not write this as a text_pyharfbuzz provider?

HarfBuzz is not capable of being a text provider. HarfBuzz only does shaping. as I know a text provider in Kivy has to render a text, something like PIL/Cairo/SDL/... . HarfBuzz doesn't do rendering. but yes , we can implement it inside base provider.

Let me make myself clear by saying "after weeks working on this repository i found out pango was better choice".
According to behdad's quote ( creator of HarfBuzz and maintainer of Pango ):
> Pango's high-level API consists of the ```PangoLayout``` object, aka "here's a piece of text render it in this box I don't care what you do."

This functionality attracted me. because in PangoLayout the Pango could do the shaping+itemise+breaking and Cairo( I guess) could do the rendering. So we could have a Pango text provider. Harfbuzz+fribidi are too low level libraries. you need to implement many things yourself to have a stable layout.

***** of course I have never used and read about PangoLayout to see if I can implement it on kivy or not *****

The thing i was going to do was get help from HarfBuzz to give me direction and size of each word to move the cursor and let fribidi reshape the text in . and give the result to a text provider(pygame/PIL/SDL or w/e ) to render it on screen. ( the thing you did with Pango i think )
But I need some time to restudy about these things.

**Comment by ghost** [comment]
> HarfBuzz is not capable of being a text provider

Sorry, I meant hb+ft2 as a provider, that should be enough I think..? But you are right of course, getting it polished is a "long term goal"...

> This functionality attracted me. because in PangoLayout the Pango could 
> do the shaping+itemise+breaking and Cairo( I guess) could do the rendering.

Totally agree, unfortunately it was not easy to take full advantage of pango's power in kivy. In the current implementation, pango itemization is only done to "small chunks" of text, when it spans exactly one line and no markup is used. For example, markup `Hello [b]World[/b]` will result in *two render calls* to the text provider, meaning two separate pango layout+itemization runs. Same for `"Hello\nWorld"`, the words are rendered separately and then blitted to one bitmap for label texture... In a textinput, itemization is per line since it creates single-line labels under the hood.

Current PR status:

* Shaping - works (Pango), [Edit] Or maybe not... it's not clear to me if kivy smashs glyphs in the layout process. I don't think so, but could *easily* be wrong ....
* Itemize - limited, as described above
* Breaking - pango not used (it only sees single-line strings)
* Alignment - pango not used (works for label in pr)

Next step will be to try implementing halign for textinput, and see how it turns out.. After that I will put together a status update + build instructions + demo program and post to the PR, hopefully ready for testing when you get around to it :)

**Comment by ghost** [comment]
Alignment works for textinput as well now, to some degree anyway.. demos

* [TextInput horizontal alignment demo](https://i.imgur.com/rnmeDN7.gif) and [the demo code](https://gist.github.com/bionoid/01483540f2cc767eecd6ea7c5aac797a)
* [Font fallback and alignment demo](https://i.imgur.com/iouCwsg.gif), and [the demo code](https://gist.github.com/bionoid/5cb8ca938d0e1b995fc66e1497225455)

**Comment by RevengeComing** [comment]
@bionoid GIF images you have sent amazed me ... I'll test it today. Nice job 👍

**Comment by RevengeComing** [comment]
@bionoid There's a problem in rendering, The text doesn't reshape. In your example you used an already reshaped text. you need to reshape in runtime ( even after any change ). I try to find out where the problem lies, but you can find it faster, let me know if you found it.

**Comment by ghost** [comment]
> The text doesn't reshape.

Aha, thanks for feedback! I suspected this, but not sure about implications or right way to handle it. We probably need to call [pango_shape](https://github.com/GNOME/pango/blob/master/pango/shape.c) somewhere? But I'm not sure exactly what it involves or how to integrate its output

**Comment by ghost** [comment]
I had a quick look around, and it seems like Pango does do reshaping. Is it possible that the problem is just the byte order of the text string in Kivy? Ie that it would be reversed, it would behave as you expect?

Is both label and textinput broken currently? Or is it just the textinput?

I mean if you do this:

```
Label:
    text: 'Some RTL text'
```

.. does this render correctly / as you expect, or the problem is apparent already here?

Thanks again for feedback, much appreciated :)

**Comment by RevengeComing** [comment]
> Is both label and textinput broken currently? Or is it just the textinput?

Both of them are broken.

> .. does this render correctly / as you expect, or the problem is apparent already here?

There's no problem with english characters.

let me clear you about reshaping ( in persian ).

We have letters that has multiple shapes but they are one, they only has different shapes depending on where they are.

For example our S is س. if it is alone it is س. if its at first of a word its like ﺳ . if it's between two other letters its like ﺴ . and if its at the end it's like ﺲ .

So there's 4 shape for a letter. in UTF for each shape we have a code for س :
س : uni0633
ﺳ : uniFEB3
ﺴ : uniFEB4
ﺲ : uniFEB2

when we are typing, the keyboard asks program to render the uni0633( alone shape ), but program asks HarfBuzz for its correct shape. and HarfBuzz returns the correct shape of the word.

------------------
Also i was checking more about your work, I saw the auto alignment doesn't work for me but in the GIF you sent it behaves completely ok.

-----------------
I guess ( but not completely sure )  that you don't use the reshaped text, you keep continue with the inputed text. :laughing:

**Comment by ghost** [comment]
> auto alignment doesn't work for me but in the GIF you sent it behaves completely ok.

That's weird, it just calls `pango_find_base_dir` and use the first one it encounters. This implementation is hacky for several reasons, but I have no idea why it doesn't work... could maybe be down to some system/locale issue?

It's a bit hard to understand the reshaping problem, when all of the characters are incomprehensible ;) 

Can you please adapt this to a minimal failing example?

```python
from kivy.base import runTouchApp
from kivy.lang import Builder

runTouchApp(Builder.load_string('''
BoxLayout:
    text: 'PUT PERSIAN TEXT HERE'
    Label:
        text: root.text
        text_size: self.size
    TextInput:
        text: root.text
'''))
```

Put in some Persian here that renders incorrectly, and post the code + screenshot? Then we can compare results, just to rule out issues like library versions and locale - it is weird the auto halign doesn't work for you... 

I guess in the end, this is another sign that a complete rewrite of text layout is needed for a proper solution :/

**Comment by ghost** [comment]
> I guess ( but not completely sure ) that you don't use the reshaped text, you keep continue with the inputed text.

Yes this is correct, I don't currently do anything with the bytes in `.text` property, just pass them along to Pango for rendering. [edit: remove incorrect technical details to avoid confusion for future readers]

**Comment by RevengeComing** [comment]
Code :
```
from kivy.base import runTouchApp
from kivy.lang import Builder

runTouchApp(Builder.load_string('''
BoxLayout:
    text: 'این یک نوشته ی تستی است.'
    Label:
        text: root.text
        text_size: self.size
    TextInput:
        text: root.text
'''))
```

Unfortunately firefox reshapes it when I copy+paste it here.
If you put this inside kivy's InputText or SublimeTE  you will see it will change to something else.

-------------------------
I found another example to show you, I copied them from pyharfbuzz test case, Firefox shows unshaped text as shaped and shaped one as unshaped =)) :
```
unshaped = "یک نوشته ی تستی"
shaped = "ﯽﺘﺴﺗ ی ﻪﺘﺷﻮﻧ ﮏﯾ"
```

if you put the unshaped(Here in firefox its shaped) text inside TextInput it will change to unshaped.
and if you put shaped one inside it it will be ok.

----------------------------------------------------------------
> where each items may consist of more than one byte

can you show me the output ? they might be something like:
[{'glyph_name': u'uni0633', 'x_advance': 44, 'char': u'\u0633', 'cluster': 0L, 'y_offset': 0, 'x_offset': 0, 'y_advance': 0}]

**Comment by ghost** [comment]
Thank you for the example, that is very helpful! Can you please show a screenshot of how these strings are *supposed* to look as shaped and unshaped? .. Just so I recognize the jackpot if I hit it haha. I included them directly from your tests/harfbuzz.py, and render on my system, [code](https://gist.github.com/bionoid/8fba39bd452116d7b199987fab0c28e0):

![png](https://i.imgur.com/FqV1Afb.png)

This is the same result you see in your end?

I just stumbled on this comment from [pango/pango-shape.c](https://github.com/GNOME/pango/blob/e9471764d19f740f756cf7e6d11aba0cbba3d2ae/pango/shape.c#L79-L83) which indicates that we might be able to do shaping across boundaries without completely replacing the kivy text layout, that's good news! ... This stuff is called during the layout process... not sure what's missing to make it work.

Also confused why the auto halign doesn't work for you -- can you confirm that console reports `[INFO   ] [Text        ] Provider: pango` at startup? Can you try the function that's used for detecting direction?

```
>>> from kivy.core.text import Label
>>> Label._find_base_direction('تستی')
'rtl'
>>> Label._find_base_direction('test')
'ltr'
```

If this doesn't work, `pango_find_base_dir` somehow fails....or it's not using LabelPango class - try `from kivy.core.text.text_pango import LabelPango` and repeat to make sure

> can you show me the output ?

Sorry, I misremembered this -- not sure which file/structure I saw it. Will let you know if I find back to it..

**Comment by RevengeComing** [comment]
> Also confused why the auto halign doesn't work for you -- can you confirm that console reports [INFO ] [Text ] Provider: pango at startup? Can you try the function that's used for detecting direction?

Oh my god :cry:  , I was using SDL provider whole the time.
in logs i get :
```
[DEBUG  ] [Text        ] Provider <pango> ignored by config
[INFO   ] [Text        ] Provider: sdl2(['text_pango'] ignored)
```
so i used:
```
import os
os.environ['KIVY_TEXT'] = 'pango'
```

and everything ( shaping, alignment ) went well. ( I'm sorry again )
the only thing remains is cursor position:
- when i click at the end of the text cursor goes to the end but when i click the <- button nothing happens, but when i click on the delete button it remove characters from start. and when i click the end button it goes to the first of the input and buttons behave vice verse.

Edit: If you need help for anything let me know.

**Comment by ghost** [comment]
> I was using SDL provider whole the time.

Ohh haha, well, that's good news! I was sure the pango provider would be used automatically because it's inserted first in the provider list. I'm not sure what the "ignored by config" thing is, but even on my system without that notice, unsetting `KIVY_TEXT` results in sdl2 provider. So I must've missed something - my bad, sorry for the confusion.

Anyway, it works! Cool :D 

Could you try and see if you can provoke a line breaking / boundary shaping issue? I am not sure how serious it is in practice, it would be nice to have a data point. There are two cases,

1. Long line of text that doesn't contain any whitespace
    * Kivy will measure the rendered width of N bytes of text, incrementing N until it exceeds the available horizontal space. So if any shaping bytes change the *width* of a glyph, Kivy could chop the text here and leave other shaping bytes dangling after the line break.
2. Using markup
    * `Hello[b]world[/b]`, if the 'o' and 'w' affect each others shape, `[b]` prevent pango from seeing both at the same time. Does this cause cause issues with normal use?

These issues probably can't be fixed at this stage, but it would be nice to know if they have practical impact. Basically the question is how big to make the warning :)

Regarding cursor, I think some of the buttons should be flipped in RTL maybe? But since we don't actually know about boundaries in the text, the solution will probably not be very good. I'm hoping we can find some hacks that make it at least mostly usable as a start..

**Comment by RevengeComing** [comment]
> 1.Long line of text that doesn't contain any whitespace 

Kivy expect each character to have same width( depends on font_size ) but some characters in persian has less width than expected and the calculation goes wrong : it goes to next line before it reachs the end of Widget.

It is solvable since both HarfBuzz and Pango implements functionality to get width of the text. I think you should use these functions instead of what is already being use in Kivy.

> 2.Using markup

Its completely OK.

> Regarding cursor, I think some of the buttons should be flipped in RTL maybe?

The whole calculation of cursor should be re implemented using Pango pango_layout_get_cursor_pos or something similar.

**Comment by ghost** [comment]
> Kivy expect each character to have same width( depends on font_size ) [ ... ] it goes to next line before it reachs the end of Widget

Hmm.. Kivy does not generally assume a fixed width (kerning would not be possible, and only monospace fonts). It may be broken logic that measure size somewhere... Can you make a demo snippet? 

> I think you should use these functions instead of what is already being use in Kivy.

Pango is already used to get the width of the text, ie `pango_layout_get_pixel_size()` on a layout with the correct text/settings (this is also how wrapping is measured). I'm sure it's not correctly working everywhere though, especially not in textinput.. 

> The whole calculation of cursor should be re implemented using Pango pango_layout_get_cursor_pos

Yes absolutely, but this is not easy since the layout is not used for the full text. I think not worth it to hack this on top of existing textinput. I'd rather clean up and fix the rest of the stuff first, since it would need need to pango_itemize the full text, and then have some weirdo layer to adapt this information to the kivy text layout. I think the proper way is to refactor the core text stuff, so at the moment, I am more interested in hacks that can make it *passable* for RTL/bidi.. 

Of course I didn't try, and maybe it's not as hard/messy as I think. If you have time to research/contribute, that would be most welcome :)

**Comment by RevengeComing** [comment]
> Hmm.. Kivy does not generally assume a fixed width (kerning would not be possible, and only monospace fonts). It may be broken logic that measure size somewhere... Can you make a demo snippet?

Sorry I was testing in hurry so I miss understand the behavior :disappointed: . It breaks it into two part when its going to add line breaking.
![kivy_pango](https://user-images.githubusercontent.com/23046907/38195554-378cb8ca-3693-11e8-818e-2e501f7d403d.gif)

Its ok too. We don't have that long words ( i don't know about other languages ) .

> If you have time to research/contribute, that would be most welcome :)

I will work on cursor position, is it ok ? shall i create PR on your repository ?

**Comment by ghost** [comment]
>  It breaks it into two part when its going to add line breaking.

Do you get any pango warnings in console? It's possible that some text is not drawn due to missing clipping (pushed a temp fix for this to avoid a crash yesterday). It's not 100% clear if the provider should just clip the blit and be happy, or the text layout engine going out of bounds is a bug...

> I will work on cursor position, is it ok ? shall i create PR on your repository ?

That would be amazing! :) Ask if something is unclear of course, the code needs cleanup. I will fix the API for font_context property next I think, that's all done magically behind the scenes at the moment. 

And yes please do PR to my repo, I think this is the best way unless you have other suggestions?

**Comment by ghost** [comment]
FYI I have pushed a bunch of cleanups to `_text_pango.pyx`, should be easier to follow now :)

**Comment by RevengeComing** [comment]
@bionoid did you push the commit (cleanups) ?

**Comment by ghost** [comment]
Yes, everything has been pushed to github! Next step for me is probably to explore options for Label APIs, the current solution is very hacky with reusing `font_name` for family
